name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/csv/CSVImporter.java

```diff
@@ -290,6 +294,162 @@ public M parseObject(String source, ParsePosition pos)
         }
     }
 
+    public static class ISINField extends CSVImporter.Field
+    {
+
+        /* package */ ISINField(String name)
+        {
+            super(name);
+        }
+
+        public ISINFormat createFormat(List<Security> securityList)
+        {
+            return new ISINFormat(securityList);
+        }
+    }
+
+    public static class ISINFormat extends Format
+    {
+        private static final long serialVersionUID = 1L;
+        private List<String> isinList = new ArrayList<String>();
+        private static final String ISINpattern = "[A-Z]{2}[A-Z0-9]{9}\\d";
+
+        public ISINFormat(List<Security> securityList)
+        {
+            for (Security security : securityList)
+            {
+                if (security.getIsin() != null)
+                {
+                    String ISIN = security.getIsin().trim();
+                    if (CheckISIN(ISIN))
+                    {
+                        this.isinList.add(ISIN);
+                    }
+                }
+            }
+        }
+
+        @Override
+        public StringBuffer format(Object obj, StringBuffer toAppendTo, FieldPosition pos)
+        {
+            String s = (String)obj;
+            if (s == null)
+                throw new IllegalArgumentException();
+
+            return toAppendTo.append(s);
+        }
+
+        static boolean CheckISIN(String isin)
+        {
+            isin = isin.trim().toUpperCase();
+            if (!isin.matches("^"+ ISINpattern + "$"))
+                return false;
+            return Isin.isValid(isin);
+        }
+
+        @Override
+        public Object parseObject(String source, ParsePosition pos)
+        {
+            int parseSuccessIndex = source.length();
+            source = source.trim();
+            if (pos == null)
+                throw new NullPointerException();
+            Object rObj = null;
+
+            Pattern pattern = Pattern.compile(" (" + ISINpattern + ") ");
```

Comment:
Das Pattern wird nie positiv sein sein weil in Zeile 354 das "trim" drin ist. Damit werden die Leerzeichen ja entfernt - das Pattern hat aber Leerzeichen.

Ich glaube wir brauchen diesen Extracheck hier eigentlich gar nicht, denn ISIN#isValid prüft das ja ebenfalls ab.

Reply:
Jein, aber es ist zugegeben vielleicht nicht der richtige Platz das zu lösen...
Habe das primär mit CSVs von der IngDIBA getestet und da werden Umsätze leider mit integrierter ISIN im CSV angelegt. Ich wollte für diesen Fall auch die ISIN extrahieren können:
z.B: "Zins/Dividende  ISIN DE0007164600 SAP SE O."
Aber zugegeben, ist ein *hack*. Werde eine bessere Lösung bauen, aber nichtmehr heute.
Grüße,
cmaoling

Reply:
Ah, okay, verstanden. Beim CSV Import macht es absolut Sinn auch zu schauen ob die ISIN vielleicht nur Teil des Strings ist.